pkg/nodediscovery/nodediscovery.go

```diff
@@ -343,34 +275,25 @@ func (n *NodeDiscovery) updateLocalNode() {
 	if n.clientset.IsEnabled() {
 		// CRD IPAM endpoint restoration depends on the completion of this
 		// to avoid custom resource update conflicts.
-		n.UpdateCiliumNodeResource()
+		n.updateCiliumNodeResource(ln)
 	}
 }
 
-// UpdateLocalNode syncs the internal localNode object with the actual state of
-// the local node and publishes the corresponding updated KV store entry and/or
-// CiliumNode object
-func (n *NodeDiscovery) UpdateLocalNode() {
-	n.localNodeLock.Lock()
-	defer n.localNodeLock.Unlock()
-
-	n.fillLocalNode()
-	n.updateLocalNode()
-}
-
-// LocalNode syncs the localNode object with the information stored in the node
-// package and then returns a copy of the localNode object
-func (n *NodeDiscovery) LocalNode() *nodeTypes.Node {
-	n.localNodeLock.Lock()
-	defer n.localNodeLock.Unlock()
-
-	n.fillLocalNode()
-	return n.localNode.DeepCopy()
-}
-
 // UpdateCiliumNodeResource updates the CiliumNode resource representing the
 // local node
 func (n *NodeDiscovery) UpdateCiliumNodeResource() {
```

Comment:
Do we really need this now? I guess it's kind of used as a synchronization mechanism to "wait" for the CiliumNode to be initially created, but perhaps we could express that in another way? e.g. I'm worried this is racy due to the observing of LocalNodeStore and the `Get` here (this could call `localNodeStore.Get`, then be pre-empted by other goroutine to process newer `Node` via the subscription and then it'd continue with old copy).

Reply:
Good point. I also found that pretty strange, but I'm honestly not sure how we can refactor it, as looking at the usages it seems to be always called during IPAM initialization, which is an area I've very little knowledge about. 

But it doesn't seem to be prone to race conditions as it is executed synchronously before calling [StartDiscovery()](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/daemon/cmd/daemon.go#L975) (which in turn subscribes to node updates). In particular:

* [ClusterPool/MultiPool IPAM](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/daemon/cmd/daemon.go#L741)
* [CRD/ENI/AZure/Alibaba IPAM](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/daemon/cmd/daemon.go#L946)  -> [CRDAllocator](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/pkg/ipam/ipam.go#L136-L143)  -> [nodeStore](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/pkg/ipam/crd.go#L642) -> [UpdateCiliumNodeResource](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/pkg/ipam/crd.go#L105)
* [MultiPool IPAM](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/daemon/cmd/daemon.go#L946)  -> [newMultiPoolManager](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/pkg/ipam/ipam.go#L125-L134)  -> [UpdateCiliumNodeResource](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/pkg/ipam/multipool.go#L258)

My proposal would be to add a comment to `UpdateCiliumNodeResource` to make it clear that it should be called only before StartDiscovery is executed, and leave a more extensive refactoring to a future PR to avoid further extending the scope of this one, which is already pretty large.